python/ray/data/impl/pandas_block.py

```diff
@@ -164,6 +174,66 @@ def _empty_table() -> "pandas.DataFrame":
     def _sample(self, n_samples: int, key: "SortKeyT") -> "pandas.DataFrame":
         return self._table[[k[0] for k in key]].sample(n_samples, ignore_index=True)
 
+    def _apply_agg(
+        self, agg_fn: Callable[["pandas.Series", bool], U], on: KeyFn
+    ) -> Optional[U]:
+        """Helper providing null handling around applying an aggregation to a column."""
+        if on is not None and not isinstance(on, str):
+            raise ValueError(
+                "on must be a string or None when aggregating on Pandas blocks, but "
+                f"got: {type(on)}."
+            )
+
+        col = self._table[on]
+        try:
+            return agg_fn(col)
+        except TypeError as e:
+            # Converting an all-null column in an Arrow Table to a Pandas DataFrame
+            # column will result in an all-None column of object type, which will raise
+            # a type error when attempting to do most binary operations. We explicitly
+            # check for this type failure here so we can properly propagate a null.
+            if np.issubdtype(col.dtype, np.object_) and col.isnull().all():
```

Comment:
Could we just check beforehand?

Reply:
I'm purposefully not checking beforehand since this check is expensive: `col.isnull().all()` creates a fully copy of the column and then traverses it (2 scans). By only checking it on such a `TypeError`, we keep this expensive check off of the common critical path.